### PR TITLE
VAULT-6803: fix listener issue if using `proxy_protocol_behavior` with `deny_unauthorized` for untrusted upstream connections

### PR DESCRIPTION
**Pull request by peteski22**
### Description

This PR corrects an issue introduced when [supported was added](https://github.com/hashicorp/vault/pull/13540) for `PROXY v2` on TCP listeners using [`proxy_protocol_behavior`](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#proxy_protocol_behavior).

This issue was introduced in Vault `v1.10.0` and has impacted all subsequent versions other than `v1.16.x` (due to a [bug parsing this config setting](https://github.com/hashicorp/vault/pull/27459)).

Any listener configured to `deny_unauthorized` will still close a connection when the upstream proxy is untrusted, but it will now continue to listen for connections until a valid one is found. It does this by [returning `ErrInvalidUpstream` to the proxy library](https://github.com/peteski22/go-proxyproto/blob/main/protocol.go#L85).

**NOTE**: the issue is only present if this config setting `proxy_protocol_behavior` is enabled and set to `deny_unauthorized`.

### HashiCorp checklist

- [x] **Labels:** If this PR is the CE portion of an ENT change, and that ENT change is
  getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
  instead of the old style `backport/x.x.x` labels.
- [x] **Labels:** If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).
- [ ] ~**ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.~
- [x] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [ ] ~**RFC:** If this change has an associated RFC, please link it in the description.~
- [ ] ~**ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.~

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -542,3 +542,7 @@ require (
 )
 
 replace github.com/ma314smith/signedxml v1.1.1 => github.com/moov-io/signedxml v1.1.1
+
```
Comment (peteski22) [inline go.mod:545]: Info on [forked repo tag](https://github.com/peteski22/go-proxyproto/releases/tag/v1.0.0).

> Keep listener listening if upstream connection address is not trusted (but close that connection).
> Releases the changes requested on pires's go-proxyproto by https://github.com/pires/go-proxyproto/pull/110.

Bumped to v1.0.0 which doesn't exist as a tag on the maintainer's repo but does on the forked repo.

File: command/server/listener_test.go [hunk 2]
```diff
@@ -60,7 +61,11 @@ func testListenerImpl(t *testing.T, ln net.Listener, connFn testListenerConnFn,
 		}
 	}
 
-	server := <-serverCh
+	var server net.Conn
+	select {
+	case <-time.After(3 * time.Second):
+	case server = <-serverCh:
+	}
```
Comment (peteski22) [inline command/server/listener_test.go:68]: Added a timeout to the tests as the forked library now never stops listening for a valid connection.

File: helper/proxyutil/proxyutil.go [hunk 3]
```diff
@@ -72,7 +71,7 @@ func WrapInProxyProto(listener net.Listener, config *ProxyProtoConfig) (net.List
 					return proxyproto.IGNORE, nil
 				}
 
-				return proxyproto.REJECT, errors.New(`upstream connection not trusted proxy_protocol_behavior is "deny_unauthorized"`)
+				return proxyproto.REJECT, proxyproto.ErrInvalidUpstream
```
Comment (peteski22) [inline helper/proxyutil/proxyutil.go:74]: This is the change that keeps the listener waiting for a valid connection.

https://github.com/peteski22/go-proxyproto/blob/v1.0.0/protocol.go#L85-L88

File: go.mod [hunk 4]
```diff
@@ -542,3 +542,7 @@ require (
...
+// Support using the forked repository until https://github.com/pires/go-proxyproto/pull/110 merges
+// and is released.
+replace github.com/pires/go-proxyproto v1.0.0 => github.com/peteski22/go-proxyproto v1.0.0
```
Comment (peteski22) [inline go.mod:548]: Info on [forked repo tag](https://github.com/peteski22/go-proxyproto/releases/tag/v1.0.0).

> Keep listener listening if upstream connection address is not trusted (but close that connection).
> Releases the changes requested on pires's go-proxyproto by https://github.com/pires/go-proxyproto/pull/110.

Bumped to v1.0.0 which doesn't exist as a tag on the maintainer's repo but does on the forked repo.

**Review by VioletHynes — COMMENTED** [review]
Is it possible to add a test that fails before this change but passes afterwards? I'm not particularly familiar with this area of the code so it would especially help increase my confidence both now and in the future if we could exercise this code path and bug fix in a test

**Comment by peteski22** [comment]
> Is it possible to add a test that fails before this change but passes afterwards? I'm not particularly familiar with this area of the code so it would especially help increase my confidence both now and in the future if we could exercise this code path and bug fix in a test

I had been testing this manually using https://github.com/muhtasimmaleque/Vault-docker-listener-bug and Linux Vault binaries I'd built before/after the change.

**Comment by VioletHynes** [comment]
Sorry, I think I could have worded that better: is it possible to add an automated Go test for this bug?

**Comment by peteski22** [comment]
> Sorry, I think I could have worded that better: is it possible to add an automated Go test for this bug? That would also increase confidence if/when we move from the fork that it still works, without needing to re-do the manual testing



> Sorry, I think I could have worded that better: is it possible to add an automated Go test for this bug? That would also increase confidence if/when we move from the fork that it still works, without needing to re-do the manual testing

👍🏼 added https://github.com/hashicorp/vault/pull/27589/files#diff-b14d5cc2a3b63b6d1e6990ec3933ac2339bfe5a825ac166048387614a4b6862cR440

This fails before the change, and passes afterwards. Between this and the manual testing steps in the Jira associated with the branch we should be able to prevent a regression and reproduce any issues.

File: go.mod (hunk 4 above)
Comment (peteski22) [reply go.mod:548]: The change here can now be updated and the `replace` removed, as the PR was accepted, merged and tagged:

https://github.com/pires/go-proxyproto/releases/tag/v0.8.0

image

**Pull request merged**